src/gennylib/include/gennylib/conventions.hpp

```diff
@@ -115,24 +115,80 @@ inline bool operator==(const TimeSpec& lhs, const TimeSpec& rhs) {
 using Duration = typename TimeSpec::ValueT;
 
 /**
- * RateSpec defined as X operations per Y duration.
+ * BaseRateSpec defined as X operations per Y duration.
  */
-struct RateSpec {
-    RateSpec() = default;
-    ~RateSpec() = default;
+struct BaseRateSpec {
+    BaseRateSpec() = default;
+    ~BaseRateSpec() = default;
 
-    RateSpec(TimeSpec t, IntegerSpec i) : per{t.count()}, operations{i.value} {}
+    BaseRateSpec(TimeSpec t, IntegerSpec i) : per{t.count()}, operations{i.value} {}
 
     // Allow construction with integers for testing.
-    RateSpec(int64_t t, int64_t i) : per{t}, operations{i} {}
+    BaseRateSpec(int64_t t, int64_t i) : per{t}, operations{i} {}
     std::chrono::nanoseconds per;
     int64_t operations;
 };
 
-inline bool operator==(const RateSpec& lhs, const RateSpec& rhs) {
+inline bool operator==(const BaseRateSpec& lhs, const BaseRateSpec& rhs) {
     return (lhs.per == rhs.per) && (lhs.operations == rhs.operations);
 }
 
+/**
+ * PercentileRateSpec defined as X% of max throughput, where X is a whole number.
+ */
+struct PercentileRateSpec {
+    PercentileRateSpec() = default;
+    ~PercentileRateSpec() = default;
+
+    PercentileRateSpec(IntegerSpec i) : percent{i.value} {}
+
+    // Allow construction with integers for testing.
+    PercentileRateSpec(int64_t i) : percent{i} {}
+    int64_t percent;
+};
+
+inline bool operator==(const PercentileRateSpec& lhs, const PercentileRateSpec& rhs) {
+    return (lhs.percent == rhs.percent);
+}
+
+/**
+ * RateSpec defined as either X operations per Y duration or Z% of max throughput each phase.
+ */
+class RateSpec {
+public:
+    RateSpec() = default;
+    ~RateSpec() = default;
+
+    RateSpec(BaseRateSpec s) : _spec{s} {}
+
+    RateSpec(PercentileRateSpec s) : _spec{s} {}
+
+    std::optional<BaseRateSpec> getBaseSpec() const {
+        if (auto pval = std::get_if<BaseRateSpec>(&_spec)) {
+            return *pval;
+        } else {
+            return std::nullopt;
+        }
+    }
+
+    std::optional<PercentileRateSpec> getPercentileSpec() const {
+        if (auto pval = std::get_if<PercentileRateSpec>(&_spec)) {
+                return *pval;
+        } else {
+            return std::nullopt;
+        }
+    }
+
+    inline bool operator==(const RateSpec& rhs) {
```

Comment:
everything in a header is implicitly inline

Reply:
Fixed.